lib/galaxy/objectstore/irods.py

```diff
@@ -376,73 +429,77 @@ def _push_to_irods(self, rel_path, source_file=None, from_string=None):
-                    return True
+        # Check if the data object exists in iRODS
+        collection_path = self.home + "/" + str(subcollection_name)
+        data_object_path = collection_path + "/" + str(data_object_name)
+        exists = False
+        try:
+            exists = self.session.data_objects.exists(data_object_path)
 
-                if from_string:
-                    data_obj = self.session.data_objects.create(data_object_path, self.resource, **options)
-                    with data_obj.open('w') as data_obj_fp:
-                        data_obj_fp.write(from_string)
-                    log.debug("Pushed data from string '%s' to collection '%s'", from_string, data_object_path)
-                else:
-                    start_time = datetime.now()
-                    log.debug("Pushing cache file '%s' of size %s bytes to collection '%s'", source_file, os.path.getsize(source_file), rel_path)
+            if os.path.getsize(source_file) == 0 and exists:
+                log.debug("Wanted to push file '%s' to iRODS collection '%s' but its size is 0; skipping.", source_file, rel_path)
+                return True
 
-                    # Create sub-collection first
-                    self.session.collections.create(collection_path, recurse=True)
-                    data_obj = self.session.data_objects.create(data_object_path, self.resource, **options)
+            if from_string:
+                data_obj = self.session.data_objects.create(data_object_path, self.resource, **options)
+                with data_obj.open('w') as data_obj_fp:
+                    data_obj_fp.write(from_string)
+                log.debug("Pushed data from string '%s' to collection '%s'", from_string, data_object_path)
+            else:
+                start_time = datetime.now()
+                log.debug("Pushing cache file '%s' of size %s bytes to collection '%s'", source_file, os.path.getsize(source_file), rel_path)
 
-                    # Write to file in subcollection created above
-                    with open(source_file, 'rb') as content_file, data_obj.open('w') as data_obj_fp:
-                        for chunk in iter(partial(content_file.read, CHUNK_SIZE), b''):
-                            data_obj_fp.write(chunk)
+                # Create sub-collection first
+                self.session.collections.create(collection_path, recurse=True)
 
-                    end_time = datetime.now()
-                    log.debug("Pushed cache file '%s' to collection '%s' (%s bytes transfered in %s sec)",
-                            source_file, rel_path, os.path.getsize(source_file), end_time - start_time)
-                return True
-            except NetworkException as e:
-                log.exception(e)
-                return False
+                self.put_data_object(source_file, collection_path)
```

Comment:
Instead of directly writing the file, use this helper method that has logic to use iput for large files.